src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
 
-            //read the icon header
-            BinaryReader reader = new BinaryReader(stream);
+            bool sizeObtained = false;
+            bool valid = false;
+            // read the icon header
+            using (var reader = new BinaryReader(stream))
+            {
+                // iconDir = new IconDir ();
+                iconDir.idReserved = reader.ReadUInt16();
+                if (iconDir.idReserved != 0) //must be 0
+                    throw new System.ArgumentException("Invalid Argument", nameof(stream));
+
+                iconDir.idType = reader.ReadUInt16();
+                if (iconDir.idType != 1) //must be 1
+                    throw new System.ArgumentException("Invalid Argument", nameof(stream));
+
+                ushort dirEntryCount = reader.ReadUInt16();
+                imageData = new ImageData[dirEntryCount];
+                iconDir.idCount = dirEntryCount;
+                iconDir.idEntries = new IconDirEntry[dirEntryCount];
+
+                // now read in the IconDirEntry structures
+                for (int i = 0; i < dirEntryCount; i++)
+                {
+                    IconDirEntry ide = new IconDirEntry
+                    {
+                        width = reader.ReadByte(),
+                        height = reader.ReadByte(),
+                        colorCount = reader.ReadByte(),
+                        reserved = reader.ReadByte(),
+                        planes = reader.ReadUInt16(),
+                        bitCount = reader.ReadUInt16(),
+                        bytesInRes = reader.ReadUInt32(),
+                        imageOffset = reader.ReadUInt32()
+                    };
+                    iconDir.idEntries[i] = ide;
+
+                    long streamPosition = stream.Position;
+                    stream.Seek(iconDir.idEntries[i].imageOffset, SeekOrigin.Begin);
+                    byte[] buffer = new byte[iconDir.idEntries[i].bytesInRes];
+                    stream.Read(buffer, 0, buffer.Length);
+                    using (var bihReader = new BinaryReader(new MemoryStream(buffer)))
+                    {
+                        uint headerSize = bihReader.ReadUInt32();
 
-            //iconDir = new IconDir ();
-            iconDir.idReserved = reader.ReadUInt16();
-            if (iconDir.idReserved != 0) //must be 0
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                        // Vista 256x256 icons points directly to a PNG bitmap
+                        // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
+                        // and we ignore them just like MS does (at least up to fx 2.0)
+                        // Added: storing data so it can be saved back
+                        if ((ide.width == 0 && ide.height == 0) || headerSize == (uint)(('P' << 8) | ('N' << 16) | ('G' << 24) | 0x89))
```

Comment:
This is the real check that I introduced to validate if the entry is in PNG format.

Reply:
Seems the magic number is actually 8 bytes, should we look at it all?
```
      The first eight bytes of a PNG file always contain the following
      values:

         (decimal)              137  80  78  71  13  10  26  10
         (hexadecimal)           89  50  4e  47  0d  0a  1a  0a
         (ASCII C notation)    \211   P   N   G  \r  \n \032 \n
```
https://tools.ietf.org/html/rfc2083

In fact I see the Icon.Windows.cs does

Reply:
And Windows itself does.

Reply:
I think we can delete the check since the specs says the first 8 bytes should be specifically those values you posted above, so I think that is enough. 

I will add the comparison of the following 4 bytes against the expected values.

If it is a bitmap also the size of the header should be 40 bits, should we add a debut assert of that if it is not a PNG image?